core/include/traccc/fitting/kalman_filter/kalman_fitter.hpp

```diff
@@ -0,0 +1,202 @@
+/** TRACCC library, part of the ACTS project (R&D line)
+ *
+ * (c) 2022 CERN for the benefit of the ACTS project
+ *
+ * Mozilla Public License Version 2.0
+ */
+
+#pragma once
+
+// Project include(s).
+#include "traccc/definitions/qualifiers.hpp"
+#include "traccc/edm/track_candidate.hpp"
+#include "traccc/edm/track_parameters.hpp"
+#include "traccc/edm/track_state.hpp"
+#include "traccc/fitting/kalman_filter/kalman_actor.hpp"
+
+// detray include(s).
+#include "detray/propagator/actor_chain.hpp"
+#include "detray/propagator/actors/aborters.hpp"
+#include "detray/propagator/actors/parameter_resetter.hpp"
+#include "detray/propagator/actors/parameter_transporter.hpp"
+#include "detray/propagator/actors/pointwise_material_interactor.hpp"
+#include "detray/propagator/propagator.hpp"
+
+namespace traccc {
+
+/// Kalman fitter algorithm to fit a single track
+template <typename stepper_t, typename navigator_t>
+class kalman_fitter {
+
+    public:
+    // scalar type
+    using scalar_type = typename stepper_t::scalar_type;
+
+    // Kalman fitter configuration
+    struct config {
+        std::size_t n_iterations = 1;
+        scalar_type pathlimit = std::numeric_limits<scalar>::max();
+        scalar_type overstep_tolerance = -10 * detray::unit<scalar>::um;
+        scalar_type step_constraint = 5. * detray::unit<scalar>::mm;
+    };
+
+    // transform3 type
+    using transform3_type = typename stepper_t::transform3_type;
+
+    // Detector type
+    using detector_type = typename navigator_t::detector_type;
+
+    // Actor types
+    using aborter = detray::pathlimit_aborter;
+    using transporter = detray::parameter_transporter<transform3_type>;
+    using interactor = detray::pointwise_material_interactor<transform3_type>;
+    using fit_actor = traccc::kalman_actor<transform3_type, vecmem::vector>;
+    using resetter = detray::parameter_resetter<transform3_type>;
+
+    using actor_chain_type =
+        detray::actor_chain<std::tuple, aborter, transporter, interactor,
+                            fit_actor, resetter>;
+
+    // Propagator type
+    using propagator_type =
+        detray::propagator<stepper_t, navigator_t, actor_chain_type>;
+
+    /// Constructor with a detector
+    ///
+    /// @param det the detector object
+    kalman_fitter(const detector_type& det) : m_detector(&det) {}
+
+    /// Kalman fitter state
+    struct state {
+
+        /// State constructor
+        ///
+        /// @param track_states the vector of track states
+        state(vecmem::vector<track_state<transform3_type>>&& track_states)
+            : m_fit_actor_state(std::move(track_states)) {}
```

Comment:
As always, I'm a bit suspicious about this move operation. Is it really not meant to be possible to construct a "state object" without destroying a vector of track states in the process? :confused: The "move constructor" could be one good constructor for this type. But I could definitely imagine that a copy-constructor could also be necessary.

Though I definitely do not know how these objects are actually meant to interact with each other. :thinking:

Reply:
>  But I could definitely imagine that a copy-constructor could also be necessary.

I am not sure. It is not very required for the moment but it could be useful in the future